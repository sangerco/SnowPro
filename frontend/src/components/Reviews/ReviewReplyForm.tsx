import React, { useState } from 'react';
import { connect } from 'react-redux';
import { useAuth } from '../AuthProvider';
import { sendNewReviewReplyData } from '../../redux/actions/replyReviewActions';
import { RootState } from '../../redux/store';
import { Button, Container, Form, Message } from 'semantic-ui-react';
import { useParams } from 'react-router';

interface NewReviewReplyData {
    userId: string;
    reviewId: string;
    body: string;
}

interface NewReviewReplyProps {
    newReviewReply: NewReviewReplyData | null;
    error: string;
    sendNewReviewReplyData: ( userId: string,
                                review: string,
                                body: string) => Promise<void>;
}

const ReviewReplyForm: React.FC<NewReviewReplyProps> = ({ newReviewReply, error, sendNewReviewReplyData }) => {
    const { reviewId } = useParams<{ reviewId: string }>();
    const { userId } = useAuth();
    const initialReviewReplyState = {   userId: userId ?? '',
                                        reviewId: reviewId,
                                        body: '' };

    const [ formData, setFormData ] = useState(initialReviewReplyState);

    const handleSubmit = async (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();

        await sendNewReviewReplyData(   formData.userId,
                                        formData.reviewId ?? '',
                                        formData.body);

        setFormData(initialReviewReplyState);
    };

    const handleChange: React.ChangeEventHandler<HTMLInputElement> = (e) => {
        const { name, value } = e.target;
        setFormData({
            ...formData,
            [name]: value
        });
    };

    return (
        <Container fluid>
            <Form onSubmit={handleSubmit} error>
                <Form.TextArea>
                    <label>Reply to this review.</label>
                    <input name='body' value={formData.body} onChange={handleChange} />
                </Form.TextArea>
            <Button type='submit' color='green'>Reply</Button>
            <Message 
                error
                header={error}
                content={error}
            />
            </Form>
        </Container>
    )
};

const mapStateToProps = (state: RootState) => ({
    newReviewReply: state.newReviewReply.data,
    error: state.newReviewReply.error
});

const mapDispatchToProps = {
    sendNewReviewReplyData
};

// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-nocheck
export default connect(mapStateToProps, mapDispatchToProps)(ReviewReplyForm)