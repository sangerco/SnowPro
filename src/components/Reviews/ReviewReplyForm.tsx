import React, { useState } from "react";
import { useDispatch } from "react-redux";
import { AppDispatch } from "../../redux/store";
import { createReviewReply } from "../../redux/slices/reviewReplySlice";
import { Button, Container, Form, Header } from "semantic-ui-react";

interface ReviewReplyFormProps {
  reviewId: string;
  slug: string;
  username: string;
  userId: string;
  closeModal: () => void;
}

const ReviewReplyForm: React.FC<ReviewReplyFormProps> = ({
  reviewId,
  slug,
  username,
  userId,
  closeModal,
}) => {
  const dispatch = useDispatch<AppDispatch>();
  const id = reviewId;

  const initialReviewReplyState = {
    userId: userId!,
    username: username!,
    reviewId: id,
    slug: slug,
    body: "",
  };

  const [formData, setFormData] = useState(initialReviewReplyState);

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    dispatch(createReviewReply(formData));

    setFormData(initialReviewReplyState);
  };

  const handleChange: React.ChangeEventHandler<HTMLTextAreaElement> = (e) => {
    const { name, value } = e.target;

    setFormData({
      ...formData,
      [name]: value,
    });
  };

  return (
    <Container
      fluid
      style={{ margin: "20px", padding: "20px" }}
      data-testid="review-reply-form">
      <Header as="h3">Reply to {`${username}`}'s review:</Header>
      <Form onSubmit={handleSubmit}>
        <Form.TextArea
          placeholder="Reply to this review"
          name="body"
          value={formData.body}
          onChange={handleChange}
        />
        <Button
          size="small"
          color="green"
          type="submit"
          data-testid="submit-button">
          Reply
        </Button>
        <Button size="small" color="red" onClick={closeModal}>
          Cancel
        </Button>
      </Form>
    </Container>
  );
};

export default ReviewReplyForm;
