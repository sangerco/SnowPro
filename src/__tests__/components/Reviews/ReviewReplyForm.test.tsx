/* eslint-disable testing-library/no-node-access */
import React from "react";
import ReviewReplyForm from "../../../components/Reviews/ReviewReplyForm";
import { render, fireEvent, screen, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import nock, { pendingMocks } from "nock";

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe("test Review Reply Form", () => {
  const initialState = {
    auth: {
      data: {
        id: "11",
      },
    },
    reviews: {
      review: {
        id: "review-1",
        userId: "11",
        username: "testuser",
        skiAreaSlug: "ski-area-1",
        skiAreaName: "Ski Area 1",
        header: "test review",
        body: "test review body",
        stars: 5,
        photos: [],
        createdAt: "fake created Date",
        replies: [],
      },
      reviews: [
        {
          id: "review-1",
          userId: "11",
          username: "testuser",
          skiAreaSlug: "ski-area-1",
          skiAreaName: "Ski Area 1",
          header: "test review",
          body: "test review body",
          stars: 5,
          photos: [],
          createdAt: "fake created Date",
          replies: [],
        },
        {
          id: "review-2",
          userId: "22",
          username: "testuser2",
          skiAreaSlug: "ski-area-2",
          skiAreaName: "Ski Area 2",
          header: "test review",
          body: "test review body",
          stars: 2,
          photos: [],
          createdAt: "fake created Date",
          replies: [],
        },
      ],
    },
    skiAreas: {
      skiArea: {
        slug: "ski-area-1",
        name: "Ski Area 1",
        country: "US",
        region: "CO",
        href: "www.fakehref.com",
        location: {
          latitude: 0,
          longitude: 0,
        },
        lifts: {},
        twitter: {},
      },
      skiAreas: [
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
        {
          slug: "ski-area-1",
          name: "Ski Area 1",
          country: "US",
          region: "CO",
          href: "www.fakehref.com",
          location: {
            latitude: 0,
            longitude: 0,
          },
          lifts: {},
          twitter: {},
        },
      ],
    },
  };
  const store = mockStore(initialState);
  it("should render without crashing", () => {
    render(
      <Provider store={store}>
        <BrowserRouter>
          <ReviewReplyForm
            reviewId={"review-1"}
            slug={"ski-area-1"}
            username={"testuser"}
            userId={"11"}
            closeModal={function (): void {
              throw new Error("Function not implemented.");
            }}
          />
        </BrowserRouter>
      </Provider>
    );
  });
  it("should match snapshot", () => {
    const { asFragment } = render(
      <Provider store={store}>
        <BrowserRouter>
          <ReviewReplyForm
            reviewId={"review1"}
            slug={"ski-area-1"}
            username={"testuser"}
            userId={"11"}
            closeModal={function (): void {
              throw new Error("Function not implemented.");
            }}
          />
        </BrowserRouter>
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render a form", async () => {
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .post("/api/reviews/review-1/reply", {
        reviewId: "review-1",
      })
      .reply(201, {
        reply: {
          id: "review-reply-1",
          userId: "11",
          username: "testuser",
          reviewId: "review-1",
          skiAreaSlug: "ski-area-1",
          body: "test review reply body",
          createdAt: "fake created date",
        },
      });
    render(
      <Provider store={store}>
        <BrowserRouter>
          <ReviewReplyForm
            reviewId={"review-1"}
            slug={"ski-area-1"}
            username={"testuser"}
            userId={"11"}
            closeModal={function (): void {
              throw new Error("Function not implemented.");
            }}
          />
        </BrowserRouter>
      </Provider>
    );
    await waitFor(() => {
      const reviewReplyForm = screen.getByTestId("review-reply-form");
      expect(reviewReplyForm).toBeInTheDocument();
    });
    await waitFor(() => {
      const body = screen.getByPlaceholderText("Reply to this review");
      expect(body).toBeInTheDocument();
    });
    await waitFor(() => {
      const button = screen.getByTestId("submit-button");
      expect(button).toBeInTheDocument();
    });
  });

  it("should submit the form", async () => {
    nock("http://localhost:5000")
      .defaultReplyHeaders({
        "access-control-allow-origin": "*",
      })
      .post("/api/reviews/review-1/reply", {
        reviewId: "review-1",
      })
      .reply(201, {
        reply: {
          id: "review-reply-1",
          userId: "11",
          username: "testuser",
          reviewId: "review-1",
          skiAreaSlug: "ski-area-1",
          body: "test review reply body",
          createdAt: "fake created date",
        },
      });
    render(
      <Provider store={store}>
        <BrowserRouter>
          <ReviewReplyForm
            reviewId={"review-1"}
            slug={"ski-area-1"}
            username={"testuser"}
            userId={"11"}
            closeModal={function (): void {
              throw new Error("Function not implemented.");
            }}
          />
        </BrowserRouter>
      </Provider>
    );

    expect(nock.pendingMocks().length).toBe(1);

    await waitFor(() => {
      const button = screen.getByTestId("submit-button");
      // eslint-disable-next-line testing-library/no-wait-for-side-effects
      fireEvent.click(button);
    });
  });
});
