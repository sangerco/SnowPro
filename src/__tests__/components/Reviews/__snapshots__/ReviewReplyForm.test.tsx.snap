// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test Review Reply Form should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="ui fluid container"
    data-testid="review-reply-form"
    style="margin: 20px; padding: 20px;"
  >
    <h3
      class="ui header"
    >
      Reply to testuser's review:
    </h3>
    <form
      class="ui form"
    >
      <div
        class="field"
      >
        <textarea
          name="body"
          placeholder="Reply to this review"
          rows="3"
        />
      </div>
      <button
        class="ui green small button"
        data-testid="submit-button"
        type="submit"
      >
        Reply
      </button>
      <button
        class="ui red small button"
      >
        Cancel
      </button>
    </form>
  </div>
</DocumentFragment>
`;
